cep-plugin-infrastructure.md

```diff
@@ -0,0 +1,135 @@
+<table>
+<tr><td> Title </td><td> Implement initial conda plugin mechanism </td>
+<tr><td> Status </td><td> Draft  </td></tr>
+<tr><td> Author(s) </td><td> Bianca Henderson &lt;bhenderson@anaconda.com&gt;</td></tr>
+<tr><td> </td><td>Jannis Leidel &lt;jleidel@anaconda.com&gt; </td></tr>
+<tr><td> Created </td><td> July 5, 2022 </td></tr>
+<tr><td> Updated </td><td> July 6, 2022 </td></tr>
+<tr><td> Discussion </td><td>https://github.com/conda-incubator/ceps/pull/32</td></tr>
+<tr><td> Implementation </td><td>https://github.com/conda/conda/pull/11435</td></tr>
+</table>
+
+## Abstract
+
+In order to enable customization and extra features that are compatible with and discoverable by `conda` (but do not necessarily ship as a default part of the `conda` codebase), we would like to implement an official `conda` plugin mechanism.
+
+A plugin mechanism in `conda` would provide many benefits, including (but not limited to) the coverage of underserved use cases, the added ability to extend `conda` internals via official APIs, as well as lowering the barrier of entry for contributions from other stakeholders in the `conda` community and ecosystem.
+
+The [initial proposal to add this plugin architecture](https://github.com/conda-incubator/ceps/blob/main/cep-2.md) has been [officially approved](https://github.com/conda-incubator/ceps/issues/23); this current CEP will discuss the specific way that the plugin mechanism will actually be implemented.
+
+
+## Specification
+
+Plugins in `conda` will integrate the "hook + entry point" structure by utilizing the [`pluggy`](https://pluggy.readthedocs.io/en/stable/index.html) Python framework. This implementation can be broken down via the following two steps:
+
+- Define the hook(s) to be registered
+- Register the plugin under the `conda` entrypoint namespace
+
+
+### Hook
+
+Below is an example of a very basic plugin "hook":
```

Comment:
Do we want to define different types of hooks in this CEP? For example, I did some quick brainstorming was able to come up with the following:

| Hook                       | Description                                                              |
|----------------------------|--------------------------------------------------------------------------|
| **Sub-command hook**       | Allows plugin authors to create an entirely new sub-command as is currently described in this open pull-request: https://github.com/conda/conda/pull/11435 |
| **Exception handling hook**| Allows a plugin author to hook in to the way that exceptions are handled and open up the possibility to add telemetry for exception logging (e.g. via services like Sentry) |
| **HTTP/CondaSession hook** |Allows a plugin author to completely override the way that HTTP requests are handled in `conda`. This opens up the possibility for more complex authentication mechanisms that we currently cannot support (e.g. OAuth workflows). This too could be used for telemetry for logging. |
| **Solver hook**            | Allows plugin authors to either modify or completely override the solver used in `conda`. |
| **pre/post command hook** | Allows plugin authors to run something either before or after a list of specified commands (i.e. arbitrary command decorators) |

I would like to see us detail as many of these hooks as possible in this CEP and leave the door open for defining new hooks when we need.

If we do end up just leaving the sub-command hook in here for now as the only hook, could we at least refer to it as the "Sub-Command Hook" and announce that there will be more hooks added as necessary?

Reply:
I think we'd want to include the ones that make sense but avoid trying to create an exhaustive list. Somewhat of YAGNI at play, we could try to spend too much time thinking of future scenarios that might not come up. We know we won't be able to capture everything so I'd suggest we outline some pragmatic choices with perhaps details on how to document new hooks in the future.

Reply:
I am also skeptical about adding a lot of hooks at once. Each hook is an open door we might not be able to close in the future, and I would think very carefully about adding hooks (and where exactly in the codebase).

> could we at least refer to it as the "Sub-Command Hook" and announce that there will be more hooks added as necessary?

This approach sounds perfect to me!